RBD mirror Image feature reflection test case automation

sunilangadi2 opened the pull request:
# Description
Automated test case [CEPH-9520](https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-9520)

**Test Case Flow:**
    1. Create a pool on both clusters.
    2. Configure mirroring (peer bootstrap) between two clusters.
    3. Enable pool mode journal based mirroring on the pool respectively.
    4. Verify if change of any image features in primary site gets mirrored in secondary site.
    5. Enable image mode snapshot based mirroring on the pool respectively.
    6. Verify if change of any image features in primary site gets mirrored in secondary site.
    7. Perform both scenario on EC and Replicated pool
    
_Test suite modified:_
suites/pacific/rbd/tier-2_rbd_mirror_regression.yaml
suites/quincy/rbd/tier-2_rbd_mirror_regression.yaml

_New test module added:_
tests/rbd_mirror/test_rbd_mirror_image_features.py
    
Test result:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-98X8XP

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

VasishtaShastry commented:
@sunilangadi2 please attach logs
will update further once logs are available

manasagowri commented on suites/pacific/rbd/tier-2_rbd_mirror_regression.yaml:
```diff
@@ -283,3 +283,13 @@ tests:
       desc: Testing journal mirroring to snapshot mirroring
+
+  - test:
+        desc: Enable and disable image feture from primary gets reflected to secondary
```
Please correct the typo for the word feature.

manasagowri commented on suites/quincy/rbd/tier-2_rbd_mirror_regression.yaml:
```diff
@@ -276,3 +276,13 @@ tests:
       desc: Testing journal mirroring to snapshot mirroring
+
+  - test:
+      desc: Enable and disable image feture from primary gets reflected to secondary
```
Please correct the typo for the word feature.

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,102 @@
+    try:
+        log.info("Running RBD mirroring test case for image features reflection")
+        config = kw.get("config")
+        mirror1, mirror2 = [
```
Try using the rbd_mirror_config method from mirror_utils, so we can include ec pool inclusion as well as part of test case automation itself.
We need not work on ec pool inclusion separately.

sunilangadi2 replied:
Done

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,102 @@
+        # wait till image gets reflected to secondary
+        time.sleep(60)
+        # Verify image feature reflection in secondary cluster
+        out = rbd2.image_info(poolname, imagename)
```
Were the features object-map and fast-diff enabled by default? So that when you verify that they have been disabled its the actual result of disabling and not default status showing up

rkhambad replied:
2.11. Enabling and disabling image features
The block device images, such as fast-diff, exclusive-lock, object-map, or deep-flatten, are enabled by default.

manasagowri replied:
Makes sense. Thanks for the clarity.

manasagowri reviewed (commented):
Few basic questions. Will provide further review once we have logs.

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,102 @@
+    2. Configure mirroring (peer bootstrap) between two clusters.
+    3. Enable pool mode journal based mirroring on the pool respectively.
+    4. Verify if change of any image features in primary site gets mirrored in secondary site.
+    5. Enable image mode snapshot based mirroring on the pool respectively.
```
Did not see coverage for snapshot based mirroring in module.

sunilangadi2 replied:
Done covered for snapshot-based as well

sunilangadi2 commented:
```
[ceph: root@ceph-rbd2-sangadi-0l5d5q-node1-installer /]# rbd mirror image status rep_pool_cftloulWyt/rep_image_mvuKGFngJm
rep_image_mvuKGFngJm:
  global_id:   12f4d1d5-8efb-4b90-9ef9-e8ca9afba296
  state:       up+replaying
  description: replaying, {"bytes_per_second":424488.0,"entries_behind_primary":18435,"entries_per_second":102.53,"non_primary_position":{"entry_tid":243711,"object_number":63,"tag_tid":2},"primary_position":{"entry_tid":1,"object_number":65,"tag_tid":3},"seconds_until_synced":179}
  service:     ceph-rbd2-sangadi-0l5d5q-node5.wmnfxl on ceph-rbd2-sangadi-0l5d5q-node5
  last_update: 2023-03-07 09:15:27
  peer_sites:
    name: primary
    state: up+stopped
    description: local image is primary
    last_update: 2023-03-07 09:15:06
```
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-X1KEP2
In journal based mirroring `entries_behind_primary` are too slow because of that change of image feature in primary not getting reflected in secondary
need to figure out way to resolve this waiting time
@VasishtaShastry @manasagowri any solution how can we wait here?

VasishtaShastry commented:
> In journal based mirroring `entries_behind_primary` are too slow because of that change of image feature in primary not getting reflected in secondary need to figure out way to resolve this waiting time @VasishtaShastry @manasagowri any solution how can we wait here?

As suggested before, please reduce the io (override default values to initial_mirror_config) before operation under focus and may be after executing operation (ex: image feature modification) .

VasishtaShastry reviewed (commented):
Will revisit after planned changes.

sunilangadi2 commented:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-KEMAFV
Updated PR with addressing review comments but still some test on replicated pool getting failed, 
still debugging on it 
Any help @VasishtaShastry @rkhambad  from the logs and code?

sunilangadi2 commented:
Fixed the code as per snapshot based reflection,
success logs
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-NGRDZK

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,216 @@
+            "Running RBD journal based mirroring test case for image features reflection"
+        )
+        config = kw.get("config")
+        mirror1, mirror2 = [
```
I think you dont need to create mirror1 and mirror2 again, instead you can use directly
        mirror1 = rbd_mirror.get("mirror1")
        mirror2 = rbd_mirror.get("mirror2")

sunilangadi2 replied:
thanks Done.

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,216 @@
+            "Running RBD snapshot based mirroring test case for image features reflection"
+        )
+        config = kw.get("config")
+        mirror1, mirror2 = [
```
same as above:
        mirror1 = rbd_mirror.get("mirror1")
        mirror2 = rbd_mirror.get("mirror2")

rkhambad reviewed (commented):
changes look good.

VasishtaShastry commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
I see lot of overlapping lines in both methods created for different scenarios.
Can we please optimise and make it more readable ?

sunilangadi2 replied:
sorry but i tried my best to optimize, as `journal based` and `snapshot based` have different way to verify of change of image feature reflection in secondary, I need to accommodate code according to the ways,
if you have any suggestions to optimize, can you please point out me in the code with hints?

VasishtaShastry replied:
Sunil, I mentioned try catch block as an example, I still see good scope for optimisation in terms of duplicate lines

It would be great if we can optimise further
Another example, I see similar lines 4 times (Not same, but similar)
We could bring down 36 lines down to 10-15 lines ?

  ```
       else:
                time.sleep(10)
                timeout = timeout - 10
        if found:
            break
    if found == 0:
        log.error("Image features not enabled in secondary")
        return 1
    mirror1.clean_up(peercluster=mirror2, pools=[poolname])
    return 0
```

I can consider approving without 100% optimisation, but I think I'm looking for very basic optimisation.
(I'm okay to merge the PR without these changes)

sunilangadi2 replied:
Optimized some more code with removing `found` flag with new logic,
Code optimization is never ending process I tried my best to optimize, all the validation steps are different for each case hence rest of the code is needed,
marking as resolved.

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+        # Disable image features in primary cluster
+        mirror1.image_feature_disable(imagespec=imagespec, image_feature=image_feature)
+        # wait till image gets reflected to secondary
+        time.sleep(60)
```
please avoid hard time sleep. use 300 seconds to check feature disable ad enable.

sunilangadi2 replied:
Done

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+
+        # Enable image features in primary cluster
+        mirror1.image_feature_enable(imagespec=imagespec, image_feature=image_feature)
+        time.sleep(60)
```
same as above

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+            f"{kw.get('ceph_cluster_dict').values} has less or more clusters Than Expected(2 clusters expected)"
+        )
+        log.exception(ve)
+        return 1
```
you could `return 1` outside the except block

sunilangadi2 replied:
done

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+        mirror1.image_feature_disable(imagespec=imagespec, image_feature=image_feature)
+        mirror2.resync(imagespec=imagespec)
+        # wait till image gets reflected to secondary
+        time.sleep(60)
```
use iterator to check the image features availability

sunilangadi2 replied:
done

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+        log.info("Image features got enabled successfully in secondary")
+        return 0
+
+    except ValueError as ve:
```
do not add try-except here in the functions. please take it down to run method.

sunilangadi2 replied:
done

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,211 @@
+    # To test against pool based journal mirror mode
+    if mirror_obj:
+        log.info("Executing test on replicated pool")
+        if test_journal_image_feature(
```
use a iterator to run `test_journal_image_feature` and `test_snapshot_image_feature`.

sunilangadi2 replied:
done

manasagowri replied:
I do not see any iterator being used to run these, @sunilangadi2 when you say done what are the changes you did for this comment?
cc: @sunilkumarn417

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,223 @@
+        for word in features:
+            if word not in out["features"]:
+                log.info(f"Image feature {word} got disabled successfully in secondary")
+                found = 1
```
return 1 at this point instead of found flag.

sunilangadi2 replied:
Removed the `found` flag as suggested by you, it's done.

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,223 @@
+        for word in features:
+            if word in out["features"]:
+                log.info(f"Image feature {word} enabled in secondary site")
+                found = 1
```
return 1 at this point without using flag.

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,223 @@
+        for word in search_pattern:
+            if word in out["flags"]:
+                log.info(f"Image feature {word} got disabled successfully in secondary")
+                found = 1
```
same as above

sunilkumarn417 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,223 @@
+        for word in search_pattern:
+            if word not in out["flags"]:
+                log.info(f"Image feature {word} enabled in secondary site")
+                found = 1
```
same as above

sunilangadi2 commented:
Addressed all review comments,
Test result for the new code changes:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-98X8XP
Please review it.

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,213 @@
+        if not [word for word in features if word in out["features"]]:
+            log.info("Image features got disabled successfully in secondary")
+            break
+        time.sleep(10)
```
Done

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,213 @@
+            log.info("Image features got enabled successfully in secondary")
+            break
+        time.sleep(10)
+        timeout = timeout - 10
```
Done

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,213 @@
+            log.info("Image features got disabled successfully in secondary")
+            break
+        time.sleep(10)
+        timeout = timeout - 10
```
Done

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_image_features.py:
```diff
@@ -0,0 +1,213 @@
+    # Verify image feature reflection in secondary cluster
+    while timeout:
+        # added this sleep for snapshot images not reflected yet to get info as well
+        time.sleep(60)
```
yes removed instead of `resync` i should perform `schedule ` the snapshot 
After doing that things are working fine.

sunilangadi2 commented:
Updated PR with addressing suggested comments
Test results:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-EI1L0X
@sunilkumarn417 @VasishtaShastry Please review

rkhambad reviewed (approved):
LGTM
